docs/contributing/testing.md

```diff
@@ -0,0 +1,114 @@
+
+Check that:
+
+- the examples in the review app look right and behave correctly with your changes
+- your design is consistent with the rest of GOV.UK Frontend
+
+### Add an example
+
+You should add an example to the review app if the existing examples do not reflect the changes you've made.
+
+1. Open `src/govuk/components/<COMPONENT>/<COMPONENT>.yaml`, where `<COMPONENT>` is the component you've changed.
+2. Add or update examples in the `examples` list.
+
+If you've created a new component, create a new `src/govuk/<COMPONENT>/<COMPONENT>.yaml` file instead, where `<COMPONENT>` is the name of the component you've created.
+
+## 3. Test in supported browsers and assistive technology
+
+You should test that your contribution works:
+
+- in [recommended browsers](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices#browsers-to-test-in)
+- with [recommended assistive technologies](https://www.gov.uk/service-manual/technology/testing-with-assistive-technologies#what-to-test)
+- in [Internet Explorer 8](/docs/installation/supporting-internet-explorer-8.md), 9 and 10 - components do not need to look perfect
+- when your users [override colours in Windows, Firefox and Chrome](https://accessibility.blog.gov.uk/2018/08/01/supporting-users-who-change-colours-on-gov-uk/)
+
+## 4. Run the automated tests
+
+In your project folder, run `npm test` to run the automated tests, including linting.
+
+If a test fails, you should check your code for any errors, then fix any tests you need to.
+
+## 5. Write new tests
+
+You should write new tests if you’ve created a new component, or changed the way a component works by:
+
+- changing or adding to its JavaScript code
+- changing or adding to its Nunjucks macro
+- creating or updating a Sass mixin or function
+
+If you’re not sure how to write new tests, ask a developer on your team to help you or [contact us](https://design-system.service.gov.uk/#support).
+
+### If you created a component
+
+Create the following files in the `src/govuk/components` folder:
+
+- `<COMPONENT>/<COMPONENT>.test.js` - to test functionality if the component uses JavaScript
+- `<COMPONENT>/template.test.js` - to test the Nunjucks macro
+
+Where `<COMPONENT>` is the name of the component you created.
+
+You can use the existing files in the `src/govuk/components` folder as templates for your new files.
+
+### If you changed or added to a component
+
+In the `src/govuk/components` folder, update or add tests to:
+
+- `<COMPONENT>/<COMPONENT>.test.js` - if you updated functionality
+- `<COMPONENT>/template.test.js` - if you updated the Nunjucks macro
+
+Where `<COMPONENT>` is the name of the component you changed or added to.
+
+### If you created or updated a Sass mixin or function
+
+Update or add tests in `src/govuk/<LAYER>/<LAYER>.test.js`, where `<LAYER>` is the [layer](https://github.com/alphagov/govuk-frontend/blob/master/src/govuk/README.md#structure) you’re creating or updating a mixin for].
```

Comment:
This isn't quite right. The test should be added in a file with the same 'name' as the file where the mixin or function lives. In the helpers layer, for example, we have:

```
_all.scss
_clearfix.scss
_colour.scss
_device-pixels.scss
_focused.scss
_font-faces.scss
_grid.scss
_links.scss
_media-queries.scss
_shape-arrow.scss
_spacing.scss
_typography.scss
_visually-hidden.scss
colour.test.js – tests the helpers in _colour.scss
grid.test.js – tests the helpers in _grid.scss
helpers.test.js – tests that the *entire layer* compiles and does not output CSS
media-queries.test.js – tests the helpers in _media-queries.scss
spacing.test.js – tests the helpers in _spacing.scss
typography.test.js – tests the helpers in _typography.scss
```

Reply:
Ah good to know, thanks. Maybe we're trying to be too specific here - I think your comment actually works pretty well as content? So something like:

Update or add tests in the `.test.js` file that matches the name of the `.scss` file you created or updated.
For example, if you've updated `src/govuk/helpers/_colour.scss`, update or add tests in `src/govuk/helpers/_colour.test.js`.

Reply:
Yep, though that file might not exist (for example, we have no tests for _links.scss at the minute) – which may also be fine.

Reply:
Could we say "Create the file if it does not exist?"

Reply:
Yep!